### PR TITLE
Upgrade old Stripe Connect payment methods to Stripe SCA and remove Stripe Connect

### LINKED ISSUE #7872: Remove Stripe Connect
### What we should change and why (this is tech debt)

Now that we have moved to Stripe SCA we wish to remove Stripe Connect from the codebase to simplify and clean up our code.

All Stripe Connect code is be removed
No existing payment methods should reference this old payment method.
Any reports that reference this old payment method (Enterprise Fees report?) will now reference StripeSCA

### Context

All Stripe Support teams have handled the roll out of Stripe SCA so we can boldly make code changes and migrations here.

Another payment method was removed recently so this might be useful background:
#6231
and this resulting (and still unsolved) error is a mistake we should not repeat:
#7799

### Impact and timeline

While all users have been moved from Stripe Connect it is still possible for users to have chosen this payment method and be using it. Therefore we need to ensure we do a migration of any Stripe Connect payment methods to Stripe SCA.

### LINKED ISSUE #8169: Migrated StripeConnect payment methods have empty "Account Owner"
## Description

Follow-up from  #8064.
When killing stripe Connect, we were left with one missing bit, after the migration:  **Account owner** is left empty, which breaks payments, namely in Subscriptions.

## Expected Behavior

No StripeSCA payment methods should have `Account owner` empty

## Actual Behaviour

Some StripeSCA payment methods have `Account owner` left empty

## Steps to Reproduce

Find the payment methods which have no `Account owner` attributed -> I guess this should be possible through the console

## Animated Gif/Screenshot

image

## Severity

Breaks subscriptions and potentially shopfront in backoffice payments.

### PR DESCRIPTION
**Pull request by mkllnk**
#### What? Why?

Closes #7872 and closes #8169.

We tried to remove StripeConnect in the past (#7978, #8064, #8184, #8247) but it turned out to be quite tricky. Stripe payment methods have one attribute stored as Spree Preference which is in a generic key-value store table. It makes it difficult to connect relations when changing the type of a payment method and the table contains stale data.

The main piece of this pull request is the migration in the first commit. I added specs for all problematic scenarios I found. And I found the spec very helpful to find edge cases, including stale cache bugs.

##### Before staging

- Make sure that you have several payment methods including StripeConnect (which needs to be enabled via the `StripeConnect` feature toggle).
- Change a StripeConnect payment method to StripeSCA in the admin interface and then change the owner.
- Change a StripeSCA payment method to StripeConnect and then change the owner.

##### After staging

- All StripeConnect methods should now be StripeSCA methods.
- All payment methods should keep their current owner and not revert back to a previous owner or have no owner.
- You can't create new StripeConnect methods, independent of any feature toggles.
- All reports showing payment methods should still work and not crash on trying to load an old payment method.

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

File: spec/migrations/convert_stripe_connect_to_stripe_sca_spec.rb [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    # of the current payment method, not the intermediately changed one.
+    old_stripe_connect.update_columns(type: "Spree::Gateway::StripeSCA")
+    changed_method = Spree::PaymentMethod.find(old_stripe_connect.id)
+    changed_method.preferred_enterprise_id = owner.id
+    old_stripe_connect.update_columns(type: "Spree::Gateway::StripeConnect")
+    old_stripe_connect.preferred_enterprise_id = new_owner.id
+
+    subject.up
+
+    expect(result.preferred_enterprise_id).to eq new_owner.id
```
Comment (mkllnk) [inline spec/migrations/convert_stripe_connect_to_stripe_sca_spec.rb:81]: Without clearing the cache in the migration, this spec fails because we are accessing the Stripe SCA owner with an old value (`owner` instead of `new_owner`) in the setup above. So this spec should cover stale cache issues.

**Review by apricot12 — APPROVED** [review]
This looks great. The migration seems to be much more complicated than we anticipated at first. TDD really is 🙌.

File: db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    upgrade_stripe_payment_methods
+    update_payment_method_preferences
+  end
+
+  private
+
+  # When changing the type of a payment method, we leave orphaned records in
+  # the spree_preferences table. The key of the preference contains the type
+  # of the payment method and therefore changing the type disconnects the
+  # preference.
```
Comment (Matt-Yorkley) [inline db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb:24]: That sounds like it could cause nasty issues. Is this orphaned-preferences thing something that needs cleaning up (in general, not in this PR)?

**Review by Matt-Yorkley — APPROVED** [review]
Champion :muscle:

This looks like it was really tough going. `Spree::Preferences` really don't play nicely with data migrations :sweat:

File: db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb (hunk 2 above)
Comment (mkllnk) [reply db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb:24]: This is the way the Admin UX works. It makes sure though to always override stale values when needed. So we don't think that we have nasty issues at the moment. But in general, it's not good to change the type of a polymorphic model with spree preferences. The better way would be to not allow that and create a new payment method even though that would have made the current Stripe migration more difficult.

So basically, let's phase out Spree Preferences, I'd say. Not in this PR. :smile:

Comment (mkllnk) [reply db/migrate/20220118053107_convert_stripe_connect_to_stripe_sca.rb:24]: Just for reference:
https://github.com/openfoodfoundation/openfoodnetwork/blob/b86e2c8c16252679f743bc19d02018c013f9e406/app/controllers/spree/admin/payment_methods_controller.rb#L34-L47

The type change is done first which leaves orphaned data. Then we load the new class and update with all the params which should override any stale data that would be conflicting (if you changed the type and then back again). So the only problem we have is orphaned data in the table.

**Comment by filipefurtad0** [comment]
After staging:

```
openfoodnetwork=> select count(*) from spree_payment_methods where type = 'Spree::Gateway::StripeSCA';
 count 
-------
    93
(1 row)

openfoodnetwork=> select count(*) from spree_payment_methods where type = 'Spree::Gateway::StripeConnect';
 count 
-------
     0
(1 row)
```

:tada: 


_Further testing notes_

- All StripeConnect methods should now be StripeSCA methods. - OK, as seen in the DB and UI
- All payment methods should keep their current owner and not revert back to a previous owner or have no owner - OK, owners were not changed before staging and were kept after staging
- You can't create new StripeConnect methods, independent of any feature toggles - OK, Stripe for Deletion in not found on the dropdown :+1: 
-  All reports showing payment methods should still work and not crash on trying to load an old payment method.
  - checked `/admin/reports/payments` and `admin/reports/order_cycle_management` - OK

Awesome! :rocket:

**Comment by filipefurtad0** [comment]
This also closes #8169 :tada: 
```

openfoodnetwork=> select count(*) from spree_payment_methods where type = NULL;
 count 
-------
     0
(1 row)

```

**Comment by mkllnk** [comment]
Thank you @filipefurtad0!

That testing instruction was my mistake. I forgot that you can't change the owner through the UX. Maybe there would have been a way by making it cash first, changing the enterprise and then making it a Stripe method again choosing the new enterprise but it doesn't matter. I covered that with specs and that scenario is extremely unlikely, especially because everybody migrated anyway.
